lib/plsql/package.rb

```diff
@@ -2,28 +2,34 @@ module PLSQL
 
   module PackageClassMethods #:nodoc:
     def find(schema, package)
-      if schema.select_first(
-            "SELECT object_name FROM all_objects
-            WHERE owner = :owner
-              AND object_name = :package
-              AND object_type = 'PACKAGE'",
-            schema.schema_name, package.to_s.upcase)
-        new(schema, package)
-      # search for synonym
-      elsif (row = schema.select_first(
-            "SELECT o.owner, o.object_name
-            FROM all_synonyms s, all_objects o
-            WHERE s.owner IN (:owner, 'PUBLIC')
-              AND s.synonym_name = :synonym_name
-              AND o.owner = s.table_owner
-              AND o.object_name = s.table_name
-              AND o.object_type = 'PACKAGE'
-            ORDER BY DECODE(s.owner, 'PUBLIC', 1, 0)",
-            schema.schema_name, package.to_s.upcase))
-        new(schema, row[1], row[0])
-      else
-        nil
-      end
+      package_name = package.to_s.upcase
+      find_in_schema(schema, package_name) || find_by_synonym(schema, package_name)
+    end
+
+    def find_in_schema(schema, package_name)
+      row = schema.select_first(<<-SQL, schema.schema_name, package_name)
```

Comment:
@flash-gordon Thanks, I'll review it over the weekend - too tired to think clear now.